Automattic/wp-calypso
# [WIP]: Generic WpcomProxyControls

#### Changes proposed in this Pull Request

* Create generic controls rather than ones highly tailored to that store
* Use basic generic control actions to build more complex actions
* Dealing with client credentials moved out of generic controls and into actions
  * Not every request requires `client_id` and `client_secret`
  * This means getting rid of the `createControls()` factory functions and having `createActions()` functions instead. This is why the diff is so big, but most of the actions are unchanged.

I don't mind having the factory functions too much, but another option is to stash the credentials in the store itself. That felt a little weird to me (having a `SET_CREDENTIALS` action that could be dispatched at any time seemed unnatural), but I can be convinced otherwise :)

We now have the following controls:

* `wpcomRequest()`
  * Generic way of making requests through the proxy iframe
* `fetchAndParse()`
  * Wraps a standard `fetch()` call. We couldn't juse use the `apiFetch()` action from `@wordpress/data-controls` because the later has a bunch of middleware that adds unwanted parameters to the requests we make.
* `remoteLoginUser()`
  * Only applicable to `auth` store
* `reloadProxy()`
  * Only applicable to `auth` store

#### Testing Instructions

Generally we need to test that network requests are working as expected.

* Type some different terms into the "my site is about" field
  * Suggestion box should open and show results
* Type some different terms into the domain picker
  * Domain suggestions from the backend should appear in the picker
* Go through the entire gutenboarding process while logged out
  * Passwordless signup form should work
  * Should recieve an email in your inbox
  * Site should be successfully created and navigation to the block editor works
* Go through the entire gutenboarding process while logged in
  * No signup form should appear (tests that the `getCurrentUser()` resolver works)
* Auth store working
  * Checkout this branch and in the console of `calypso.localhost` try:
    * Logging in with existing passwordless account
      * `wp.auth.submitUsernameOrEmail('validemail')`
      * Check that email arrives in inbox
    * Logging in with existing passwordful account
      * `wp.auth.submitUsernameOrEmail('validusername')`
      * `wp.auth.submitPassword('correctpassword')`
      * `wp.data.dispatch('automattic/site').createSite({blog_name: 'title'})` to test that you're immediately able to create a site after logging in.
